### PR TITLE
'istioctl experimental dashboard' command to show add-ons and sidecars

### PR DESCRIPTION
**Pull request by esnible**
Create convenience commands to access the web UIs that are installed with Istio.

This is the command I demonstrated at the UX workgroup meeting on 3/19.

### CODE REVIEW

**Comment by douglas-reid** [comment]
Any concerns around using `dashboard`?  We have the grafana dashboards, which are commonly called referred to as the "Istio dashboards".  And, we have `istio experimental metrics` which provides a CLI dashboard.

**Comment by esnible** [comment]
@douglas-reid I am open to changing the command name.

My hope is that the top level `istioctl experimental dashboard` one day becomes a full featured dashboard, and the subcommands are just a more direct way to get there.

**Comment by esnible** [comment]
@geeknoid I put this PR on hold because of @douglas-reid 's question.  Is it reasonable to have users do `istioctl experimental dashboard grafana` to see the Istio's Grafana dashboard?  It makes sense to me, but what does the community think?

**Comment by geeknoid** [comment]
I feel comfortable with the use of dashboard here.

To @douglas-reid 's point though, we do have references to Istio Dashboard in the docs, primarily in the (surprise) using-istio-dashboard task. Mostly though, we use Grafana Dashboard, Zipkin Dashboard, etc.

So how about we go with this change here, and clean up the docs so that we systematically refer to the Granana stuff as the Grafana Dashboard.

File: istioctl/cmd/istioctl/dashboard.go [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+	"istio.io/istio/pkg/log"
+)
+
+var (
+	dashboardCmd = &cobra.Command{
+		Use:     "dashboard",
+		Aliases: []string{"dash", "d"},
+		Short:   "Access to Istio web UIs",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// cmd.Println(cmd.UsageString())
```
Comment (liamawhite) [inline istioctl/cmd/istioctl/dashboard.go]: tidy up?

File: istioctl/cmd/istioctl/dashboard.go [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+			case err := <-errCh:
+				return fmt.Errorf("failure running port forward process: %v", err)
+			case <-readyCh:
+				log.Debugf("port-forward to Prometheus pod ready")
+				defer fw.Close()
+
+				openBrowser(fmt.Sprintf("http://localhost:%d", port), cmd.OutOrStdout())
+
+				// Block forever
+				<-(chan int)(nil)
```
Comment (liamawhite) [inline istioctl/cmd/istioctl/dashboard.go]: I don't think this go routine is tidied up?

File: istioctl/cmd/istioctl/dashboard.go [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+	v1 "k8s.io/api/core/v1"
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/istio/pkg/log"
+)
+
+var (
+	dashboardCmd = &cobra.Command{
+		Use:     "dashboard",
```
Comment (liamawhite) [inline istioctl/cmd/istioctl/dashboard.go]: pass multiple dashboard types?

Comment (liamawhite) [reply istioctl/cmd/istioctl/dashboard.go]: could we also have an all value that just opens up every available dashboard?

File: istioctl/cmd/istioctl/dashboard.go [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+				<-(chan int)(nil)
+			}
+			return nil
+		},
+	}
+
+	return cmd
+}
+
+func podsForSelector(client cache.Getter, labelSelector string) (*v1.PodList, error) {
```
Comment (liamawhite) [inline istioctl/cmd/istioctl/dashboard.go]: move this into `istioctl/pkg` and add some testing

File: istioctl/cmd/istioctl/metrics.go [hunk 5]
```diff
@@ -161,8 +162,9 @@ func prometheusPods(client cache.Getter) (*v1.PodList, error) {
 	return obj.(*v1.PodList), nil
 }
 
-func buildPortForwarder(client *kubernetes.Client, config *rest.Config, podName string, port int) (*portforward.PortForwarder, <-chan struct{}, error) {
-	req := client.Post().Resource("pods").Namespace(istioNamespace).Name(podName).SubResource("portforward")
+// nolint: lll
+func buildPortForwarder(client *kubernetes.Client, config *rest.Config, podName string, ns string, port int, podPort int) (*portforward.PortForwarder, <-chan struct{}, error) {
```
Comment (liamawhite) [inline istioctl/cmd/istioctl/metrics.go]: Can we move this into `istioctl/pkg` and add some testing?

File: istioctl/cmd/istioctl/dashboard.go (hunk 2 above)
Comment (esnible) [reply istioctl/cmd/istioctl/dashboard.go]: I am unclear what you are suggesting.  This go routine never terminates.  Please advise.

File: istioctl/cmd/istioctl/dashboard.go (hunk 3 above)
Comment (esnible) [reply istioctl/cmd/istioctl/dashboard.go]: We have five "global" dashboard types: grafana, kiali, prometheus, jaeger, and zipkin.  A user can't install both zipkin and jaeger.  Except during a demo I probably know which dashboard type I need.

We have two "pod" dashboards.  envoy (only pods with sidecars) and controlz (a few control pane pods) that could not be part of "all".

File: istioctl/cmd/istioctl/metrics.go (hunk 5 above)
Comment (esnible) [reply istioctl/cmd/istioctl/metrics.go]: I will move it into _istioctl/pkg/kubernetes_.  There is currently no testing on that package.  I am concerned about creating testing for port-forward.  How about just the move?

**Comment by esnible** [comment]
@liamawhite I made the refactorings you requested.  Can you re-review?

File: istioctl/cmd/istioctl/dashboard.go (hunk 3 above)
Comment (liamawhite) [reply istioctl/cmd/istioctl/dashboard.go]: We could maybe use the word available? We should be able to work that out using Kube API server calls.

File: istioctl/cmd/istioctl/dashboard.go (hunk 2 above)
Comment (liamawhite) [reply istioctl/cmd/istioctl/dashboard.go]: Won't it leak the goroutine? There's no way to terminate this routine, right?

**Comment by liamawhite** [comment]
@esnible have made a couple more comments. Looks good though.

File: istioctl/cmd/istioctl/dashboard.go (hunk 3 above)
Comment (esnible) [reply istioctl/cmd/istioctl/dashboard.go]: Let's get some experience with single types.  I would like to do a follow-on later for integrating the various 3rd party dashboards.

File: istioctl/cmd/istioctl/dashboard.go (hunk 2 above)
Comment (esnible) [reply istioctl/cmd/istioctl/dashboard.go]: I made further improvements to the code hoping to get your positive review.  At the same time you `/approve`d it!  Take another look, let me know.

**Comment by esnible** [comment]
@liamawhite @ayj @fisherxu Looking for a review...

**Pull request merged**